### PR TITLE
Replace `NestedVisitorMap` with generic `NestedFilter`

### PR DESCRIPTION
**Pull request by camsteffen**
This is an attempt to make the `intravisit::Visitor` API simpler and "more const" with regard to nested visiting.

With this change, `intravisit::Visitor` does not visit nested things by default, unless you specify `type NestedFilter = nested_filter::OnlyBodies` (or `All`). `nested_visit_map` returns `Self::Map` instead of `NestedVisitorMap<Self::Map>`. It panics by default (unreachable if `type NestedFilter` is omitted).

One somewhat trixty thing here is that `nested_filter::{OnlyBodies, All}` live in `rustc_middle` so that they may have `type Map = map::Map` and so that `impl Visitor`s never need to specify `type Map` - it has a default of `Self::NestedFilter::Map`.

### CODE REVIEW

**Comment by camsteffen** [comment]
Whoops I meant to have 0 clippy changes until this gets a green light. Reverted the clippy change for now.

**Comment by cjgillot** [comment]
Seems like a great idea!
@bors try @rust-timer queue

**Comment by camsteffen** [comment]
We can't do a perf run without fixing Clippy?

**Comment by cjgillot** [comment]
I don't think so, sadly.

**Comment by camsteffen** [comment]
shucks

**Comment by Mark-Simulacrum** [comment]
You can, just needs some manual adjustment to bootstrap - probably enough to comment Clippy out everywhere in src/bootstrap/dist.rs, roughly.

**Comment by matthiaskrgr** [comment]
>We can't do a perf run without fixing Clippy?

Maybe if you add a commit that removes most of clippy (remove the tests, make it an empty `fn main() {}` project, so that clippy tests are still run but they are all empty?

File: compiler/rustc_middle/src/hir/intravisit.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+pub mod nested_filter {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/hir/intravisit.rs:1]: Why a nested module? Can't the file be directly called `nested_filter.rs`?

File: compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs [hunk 2]
```diff
@@ -447,8 +447,7 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
     /// generics:
     ///
     /// ```text
-    /// type Map = HashMap<String>;
-    /// ```
+    ///    /// ```
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs:450]: Spurious change?

File: compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs [hunk 3]
```diff
@@ -600,8 +599,7 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
     /// Suggests to remove redundant argument(s):
     ///
     /// ```text
-    /// type Map = HashMap<String, String, String, String>;
-    /// ```
+    ///    /// ```
```
Comment (cjgillot) [inline compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs:602]: Spurious change?

**Review by cjgillot — APPROVED** [review]
Thanks @camsteffen!
I left a few nits, I'll let you judge if they need fixing.
Either way, r=me after a rebase.

File: compiler/rustc_middle/src/hir/intravisit.rs (hunk 1 above)
Comment (camsteffen) [reply compiler/rustc_middle/src/hir/intravisit.rs:1]: So it is `hir::nested_filter` instead of `hir::intravisit::nested_filter`? Yeah that would be fine.

**Comment by flip1995** [comment]
As a Clippy team member, you [should be able to do so](https://github.com/rust-lang/team/blob/0aa0a5490d0e1b68019ee4d24cc0d086ec611c0f/teams/clippy.toml#L24). But apparently it didn't work.

**Comment by camsteffen** [comment]
lol

**Comment by matthiaskrgr** [comment]
Something was strange with bors, there were ~400+ PRs not listed in the queue. I have synchronized it, let's see if this helps
@bors r=cjgillot

**Comment by steffahn** [comment]
The documentation (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html) is still outdated, talking about things like “see NestedVisitorMap for details” or “override nested_visit_map to return other than None”.

**Pull request merged**